File: src/main/GO/leetcode/lists/topinterview/LeetCode212.go
package topinterview

func findWords(board [][]byte, words []string) []string {
	trie := newTrie()
	for i := 0; i < len(words); i++ {
		trie.insert(words[i])
	}
	list := make([]string, 0)
	mp := make(map[string]bool)
	visited := make([][]bool, len(board))
	for i := 0; i < len(board); i++ {
		visited[i] = make([]bool, len(board[0]))
	}
	var dfs func(root *Trie, str string, i, j int)
	dfs = func(root *Trie, str string, i, j int) {
		if i < 0 || j < 0 || i > len(board)-1 || j > len(board[0])-1 {
			return
		}
		if visited[i][j] {
			return
		}
		str = str + string(board[i][j])
		if !trie.startWith(str) {
			return
		}
		if trie.search(str) {
			mp[str] = true
			//return 不能return
		}
		visited[i][j] = true
		dfs(root, str, i+1, j)
		dfs(root, str, i, j+1)
		dfs(root, str, i-1, j)
		dfs(root, str, i, j-1)
		visited[i][j] = false
	}
	for i := 0; i < len(board); i++ {
		for j := 0; j < len(board[0]); j++ {
			dfs(trie, "", i, j)
		}
	}
	for k := range mp {
		list = append(list, k)
	}
	return list
}

File: java/src/leetcode/problems/LeetCode212.java
package leetcode.problems;



import knowledge.datastructure.trie.Trie;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * @author aiwujingxin@gmail.com
 * @date 2022/7/6 22:17
 */
public class LeetCode212 {

    Set<String> res = new HashSet<>();
    Trie trie = new Trie();

    public List<String> findWords(char[][] board, String[] words) {
        for (String word : words) {
            trie.insert(word);
        }

        int m = board.length;
        int n = board[0].length;
        boolean[][] visited = new boolean[m][n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                backtrack(board, visited, "", i, j);
            }
        }

        return new ArrayList<>(res);
    }

    public void backtrack(char[][] board, boolean[][] visited, String str, int x, int y) {
        if (x < 0 || x >= board.length || y < 0 || y >= board[0].length || visited[x][y]) {
            return;
        }
        str += board[x][y];
        if (!trie.startsWith(str)) {
            return;
        }

        if (trie.search(str)) {
            res.add(str);
        }

        visited[x][y] = true;
        backtrack(board, visited, str, x - 1, y);
        backtrack(board, visited, str, x + 1, y);
        backtrack(board, visited, str, x, y - 1);
        backtrack(board, visited, str, x, y + 1);
        visited[x][y] = false;
    }
}
